### PR TITLE
Expose the duration and curve for theme animation in MaterialApp.

### LINKED ISSUE #105883: ThemeExtension  triggers multiple rebuilds
Run the example theme_extension.1
Print content in build of the Home page. Triggering ThemeMode changes will cause Home to repeat rebuild

image

https://user-images.githubusercontent.com/17589354/173322707-3b28c8f3-fb44-4f54-b443-b50597625f35.mov

### PR DESCRIPTION
**Pull request by darrenaustin**
When a `MaterialApp`'s theme changes, it is animated over time (via the `AnimatedTheme` widget). This PR exposes the duration and curve used for this animation:

```dart
MaterialApp(
  theme: ThemeData.light(),
  darkTheme: ThemeData.dark(),
  themeMode: ThemeMode.dark,
  themeAnimationDuration: const Duration(milliseconds: 750),
  themeAnimationCurve: Curves.bounceInOut,
  // ...
```

Or if you want to turn off the animation altogether, you can set the duration to zero:

```dart
MaterialApp(
  theme: ThemeData.light(),
  darkTheme: ThemeData.dark(),
  themeMode: ThemeMode.dark,
  themeAnimationDuration: Duration.zero,
  // ...
```

Fixed: #105883

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Review by HansMuller — APPROVED** [review]
LGTM

What happens if these parameters are changed while a theme animation is underway?

File: packages/flutter/lib/src/material/app.dart [hunk 1]
```diff
@@ -472,6 +476,28 @@ class MaterialApp extends StatefulWidget {
   ///    system what kind of theme is being used.
   final ThemeMode? themeMode;
 
+  /// The duration of animated theme changes.
+  ///
+  /// When the theme changes (either by the [theme], [darkTheme] or [themeMode]
+  /// parameters changing) it is animated to the new theme over time.
+  /// [themeAnimationDuration] determines how long this animation takes.
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/app.dart:483]: nit: sentences should start grammatically correct with a capital letter.

File: packages/flutter/lib/src/material/app.dart [hunk 2]
```diff
@@ -472,6 +476,28 @@ class MaterialApp extends StatefulWidget {
...
+  ///
+  /// To have the theme change immediately, you can set this to [Duration.zero].
+  ///
+  /// The default it is [kThemeAnimationDuration].
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/app.dart:487]: nit: this sentence reads awkward. Maybe remove the "it"?

File: packages/flutter/lib/src/material/app.dart [hunk 3]
```diff
@@ -472,6 +476,28 @@ class MaterialApp extends StatefulWidget {
...
+  ///
+  /// See also:
+  ///   [themeAnimationCurve], which defines the curve used for the animation.
+  final Duration themeAnimationDuration;
+
+  /// The curve to apply when animating theme changes.
+  ///
+  /// The default it is [Curves.linear].
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/app.dart:495]: same.

File: packages/flutter/lib/src/material/app.dart [hunk 4]
```diff
@@ -472,6 +476,28 @@ class MaterialApp extends StatefulWidget {
...
+  ///
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/app.dart:496]: Maybe point out that the curve is ignored when the duration is zero?

**Review by goderbauer — APPROVED** [review]
LGTM

File: packages/flutter/lib/src/material/app.dart (hunk 2 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/app.dart]: D'oh! I english'd good. Thx.

**Comment by HansMuller** [comment]
https://github.com/flutter/flutter/pull/107383#issuecomment-1180784867 sounds OK, I don't think we need to document that case.

**Pull request merged**